### ISSUE TITLE
new Datalab issue type: underperforming group (data slice receiving bad predictions)

### ISSUE THREAD

**Issue by jwmueller** [cleanlab/cleanlab#776]
Goal is to auto-detect a coherent cluster of “hard” examples (ie. data slice) where models predictions are poor. Cf:
https://dcai.csail.mit.edu/lectures/data-centric-evaluation/

This should be a [new Datalab issue type ](https://docs.cleanlab.ai/master/cleanlab/datalab/guide/custom_issue_manager.html) called something like: `underperforming_group`.

Detecting this issue type will require these inputs: `features`, `pred_probs`, `labels`.

Steps to detect this issue: 
1. Cluster the data
2. For each cluster `c`, compute the average loss of all examples belonging to that cluster: `L_c`. Loss is computed between `pred_probs` and `labels` and can be any loss-function, say [self_confidence](https://docs.cleanlab.ai/stable/cleanlab/rank.html#cleanlab.rank.get_self_confidence_for_each_label). Let `L_w` be the the worst cluster's loss: `L_w = min_c { L_c }`
3. Compute overall prediction loss across the full dataset too: `L`.
4. If `L_w / L < c`, then we say this dataset suffers from the `underperforming_group` issue. 0 < c < 0.5 here is a hyperparameter.
5. If the dataset suffers from this issue, then we set: 
- the overall dataset quality score for the `underperforming_group` issue to be = `L_w / L` (lower values = more severe issue).
- the number of examples suffering from this issue in the dataset = the size of the worst-performing cluster.
- the quality score for each example for the `underperforming_group` issue to be = `L_w / L` for examples that belong to the worst-performing cluster, otherwise = 1.0 for other examples.

If the dataset does not suffer from this issue, then we set:
- the overall dataset quality score for the `underperforming_group` issue to be = `L_w / L` (lower values = more severe issue).
- the number of examples suffering from this issue in the dataset = 0
- the quality score for each example for the `underperforming_group` issue to be = 1.0

6. Set `Datalab.info` to keep track of other useful information computed during these steps.

**Implementation Notes**: 
Datalab already computes a KNN graph of the examples based on provided `features`. 
We should just use these pre-computed distances as the sole input to clustering algorithm. 
Recommend to start with [DBSCAN](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html#sklearn.cluster.DBSCAN) to do the clustering, using `metric='precomputed'` and passing in the already-computed KNN graph as a sparse array.

In the future, could also consider defining the clusters differently.  For instance for tabular data with categorical features, a cluster could simply be those examples which share the same value of this categorical feature (and we could iterate over all of the categorical features to find the worst-performing cluster defined based on a single one of these categorical features).

**Comment by tataganesh** [comment]
Thank you for the detailed description @jwmueller. I have three questions specific to DBSCAN

* The performance of DBSCAN depends on the `eps` hyperparameter, which should either be tuned manually, or it can be set based on some nearest neighbour heuristics (as mentioned in sklearn's [documentation](https://scikit-learn.org/stable/modules/clustering.html#dbscan)). Do you have any suggestions for assigning this hyperparameter? 
* How should we treat the samples labelled with `-1` by DBSCAN? The documentation says "noisy samples are given the label -1". 
* Should we also store metrics that evaluate the quality of clusters, like Silhouette Coefficient, as part of summary statistics?
Thanks!

**Comment by jwmueller** [comment]
@tataganesh You could actually check out:
https://scikit-learn.org/stable/modules/generated/sklearn.cluster.HDBSCAN.html 

which integrates over different values of epsilon. Just make sure this implementation is not significantly slower than DBSCAN on a large dataset (I haven't tried the sklearn HDBSCAN implementation before).

We can leave any other hyperparameters of the clustering algorithm at their default values.

We should just ignore the samples not identified as part of any cluster (i.e. -1), they would never define the worst-performing subgroup.

And yes storing some interesting outputs from the clustering would be nice, but they should be stored in:
`datalab.info["dbscan"]` or some special field of the `info` object like that which makes it clear this information is specific to the clustering-only.

**Comment by tataganesh** [comment]
Thanks @jwmueller. If we are using HDBSCAN from scikit-learn, it might be important to note that it was introduced in version 1.3. Scikit-learn's documentation says "Scikit-learn 1.1 and later requires Python 3.8 or newer." So, introducing HDBSCAN might lead to [CI tests](https://github.com/cleanlab/cleanlab/actions/runs/5942651164/job/16116060616?pr=822#step:5:24) failing for python 3.7.

**Comment by jwmueller** [comment]
> If there are only a few samples belonging to the underperforming group, but L_w / L is really low, wouldn't the overall dataset quality score be misleading here? In this case, majority of the samples do not belong to the underperforming group, so the dataset quality score should be high, right?

No the dataset quality score for this issue-type in `datalab.issue_summary` is NOT evaluating how good is the dataset overall in general, it is evaluating how severely does the dataset exhibit this particular issue overall.

In our case, a dataset that has a clear cluster with very bad model performance (low L_w / L) means this subgroup is severely underperforming. The size of the subgroup is less relevant (as long as it is big enough to define a meaningful cluster, which DBSCAN will already account for).  We are particularly interested in datasets where this cluster is quite small but the model performs very badly.  These indicate minority subgroups with disparate outcomes from model predictions.

**Comment by jwmueller** [comment]
@tataganesh  Thanks for raising the Python 3.8 issue. Don't worry about the CI failing for python 3.7, we will bump up required of version to Python 3.8 since 3.7 is well past End of Life

**Comment by tataganesh** [comment]
Thanks for the clarification @jwmueller! I realized that for class-imbalance, I updated the overall dataset quality score for the imbalance issue type to be the mean of all scores [[code](https://github.com/tataganesh/cleanlab/blob/d61988a54d09a2537de18c4a4341bb94e8fc7a35/cleanlab/datalab/internal/issue_manager/imbalance.py#L74)]. Instead, should this be just the proportion of samples in the rare class if the issue exists, and 1.0 otherwise? I wanted to clarify if I made a mistake here.

**Comment by jwmueller** [comment]
> I realized that for class-imbalance, I updated the overall dataset quality score for the imbalance issue type to be the mean of all scores

Good catch. It should've been the proportion of samples in the rare class, no matter what (whether the issue exists or not).  This allows us to quantify how severe this type of issue is overall in the dataset

**Comment by tataganesh** [comment]
Thanks! I will make the fix along with the PR for this issue.

**Comment by jwmueller** [comment]
Thank you! Could you actually push the fix in an independent PR? That should be separate from this issue

**Comment by jwmueller** [comment]
added in:
https://github.com/cleanlab/cleanlab/pull/838
https://github.com/cleanlab/cleanlab/pull/907
